### ISSUE TITLE
segfault during plv8.find_function on CentOS - PG12

### ISSUE THREAD

**Issue by nonara** [plv8/plv8js#370]
**Issue**
Segfault crash during plv8.find_function - only happens for valid functions.

**Setup**
* CentOS Linux release 7.7.1908 (Core)
* PostgreSQL 12.0 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39), 64-bit

**Detail**
* Crash occurs on both 2.3.13 and 3.0-alpha (with both clang and g++ of varying versions up to the latest)
* Seems isolated to PostgreSQL 12. (PG-11 works)

**Log**
Nov 04 18:21:29 kernel: postmaster[11344]: segfault at 0 ip 00007f120e53d252 sp 00007fff298a84d0 error 6 in plv8-3.0alpha.so[7f120e492000+c16000]
Nov 04 18:21:29 kernel: Code: 24 80 00 00 00 48 89 c1 48 83 c0 08 48 c7 40 f8 00 00 00 00 48 39 d0 75 ef 48 8b 05 a8 5b da 00 31 ff c7 44 24 58 00 00 00 00 <48> 89 0f 48 8b 00 48 89 44 24 70 b8 02 00 00 00 66 89 47 1e 89 e8

**Comment by JerrySievert** [comment]
do you have any more info or a reproducible test case?

I still have this active from another issue:

```
CREATE FUNCTION get_env(variable TEXT) RETURNS TEXT AS                                                                         $$
  switch (variable) {
    case 'current_user':  {
      return plv8.execute('SELECT current_user AS val')[0].val;
    }
  }
$$ LANGUAGE plv8;
```

```
do $$ plv8.elog(NOTICE, plv8.find_function('get_env')('current_user')); $$ language plv8;
```

which works exactly as expected.

just also making sure you compiled it with its own v8, not with something from the operating system?

**Comment by nonara** [comment]
Would be glad to provide more detail. Any suggestions on where to look?

I created a new database and added plv8 with your exact function and do command, but I encountered the same issue.

As for v8, it's being built during the make process for plv8. There isn't another on installed on the machine, so I assume it's using the correct version.

Barring isolating the issue further, the steps I believe should reproduce would be to recreate the environment, installing PostgreSQL 12 via yum on CentOS 7.7 - Not sure if that's an option for you.

**Comment by JerrySievert** [comment]
usually steps to reproduce help quite a lot.

**Comment by nonara** [comment]
Yes, of course. However, if I knew how to reproduce the issue on another OS, I likely would be able to track down the issue myself.

Presently, I've determined that it happens in 2.3.13 and 3.0-alpha, compiled with both clang and g++. This gives us a lot of information in terms of ruling things out, especially, considering the different versions of v8. 

Given the incredibly limited detail in the error log, I'm a bit stuck diagnosing it, so if you have any suggestions to further diagnose, that would be helpful.

**Comment by nonara** [comment]
Ok. **I can confirm it's reproducible.** I just spun up a fresh VM.

**Steps**
1. Spin up CentOS 7 instance (I'm using linode)
2. Install PG12  
```
yum -y update
yum -y install https://download.postgresql.org/pub/repos/yum/reporpms/EL-7-x86_64/pgdg-redhat-repo-latest.noarch.rpm
sudo yum -y install epel-release yum-utils
sudo yum-config-manager --enable pgdg12
sudo yum install postgresql12-server postgresql12 postgresql12-devel

/usr/pgsql-12/bin/postgresql-12-setup initdb
```
3. Configure PG12 so you can connect
4. Get plv8
```
yum install -y wget
cd ~
wget https://github.com/plv8/plv8/archive/v2.3.13.tar.gz
tar xvzf v2.3.13.tar.gz
cd plv8-2.3.13
```
5. Build plv8
```
yum install -y centos-release-scl
yum install -y libcxx libcxx-devel
yum install -y llvm5.0 llvm7.0 llvm7.0-devel llvm7.0-libs llvm-toolset-7
yum install -y devtoolset-8 git bzip2
scl enable devtoolset-8 bash

make PG_CONFIG=/usr/pgsql-12/bin/pg_config
make install PG_CONFIG=/usr/pgsql-12/bin/pg_config
```

6. Enable on PG (sql)
```SQL
CREATE EXTENSION plv8;
```

7. Crash (sql)
```SQL
CREATE FUNCTION get_env(variable TEXT) RETURNS TEXT AS                                                                         $$
  switch (variable) {
    case 'current_user':  {
      return plv8.execute('SELECT current_user AS val')[0].val;
    }
  }
$$ LANGUAGE plv8;

do $$ plv8.elog(NOTICE, plv8.find_function('get_env')('current_user')); $$ language plv8;
```

If you'd like any more details, let me know. Thanks!

Update: Just built Postgres 12 from source and same issue.

**Comment by nonara** [comment]
Confirmed to be working with Postgres 11 on the same machine. Looks like it's isolated to PG-12.

**Comment by cschwaderer** [comment]
Same issue for me on 

- Ubuntu 18.04
- PostgreSQL 12.1
- plv8 2.3.13

Did the same as  @nonara:

- Fresh database
- Create plv8 (`CREATE EXTENSION plv8;`)
- Create a simple plv8 function
- Use `find_function` for an existing function in a simple anonymous code block: `DO $$ plv8.find_function("my_simple_test"); $$ LANGUAGE plv8;`

Log output is

> Received signal 11 SEGV_MAPERR 000000000000
> 
> ==== C stack trace ===============================
> 
>  [0x7f7e66ddb8e5]
>  [0x7f7e7a228890]
>  [0x7f7e66607fd8]
>  [0x7f7e66607655]
>  [0x7f7e66aaab09]
>  [0x7f7e66ad877c]
>  [0x7f7e66ad7d97]
>  [0x3d195838431d]
> [end of stack trace]
> 2020-01-02 14:26:29.870 CET [16549] LOG:  server process (PID 26768) was terminated by signal 11: Segmentation fault
> 2020-01-02 14:26:29.870 CET [16549] DETAIL:  Failed process was running: DO $$
>             plv8.find_function("my_simple_test");
>         $$ LANGUAGE plv8
> 2020-01-02 14:26:29.870 CET [16549] LOG:  terminating any other active server processes
> 

On some clients, I get `SSL SYSCALL error: EOF detected` (just for reference so that issue could be found)

**Comment by cschwaderer** [comment]
@jeffcoat @JerrySievert 
May I ask what's the current state of fixing this issue? I sort of lost track in the PR thread.

This bug ultimately prevents my company from switching to PostgreSQL 12 - which is a little bit unfortunate. And I certainly don't want to wait for PLV8 3.0 to become stable before doing the update to Postgres 12.

So, is there any estimation on when a fix might reach the PLV8 2.3.X branch?

**Comment by JerrySievert** [comment]
I've back ported to 2.3.14.
